File: code/885/solution.py
# -*- coding: UTF-8 -*-
from random import random
from math import *

jiecheng = [0] * 20
cnt = [0] * 10
mod = 1123455689
ans = 0


def init():
    jiecheng[0] = 1
    for i in range(1, 20):
        jiecheng[i] = jiecheng[i - 1] * i


def C(a, b):
    if a < b:
        return 0
    return jiecheng[a] // jiecheng[b] // jiecheng[a - b]


def cal():
    s = sum(cnt)
    ret = 1
    for i in range(10):
        ret = ret * C(s - (i == 0), cnt[i])
        s -= cnt[i]
    return ret % mod


def dfs(index, need, now):
    if index == 10:
        global ans
        ans = (ans + now * cal()) % mod
        return

    buf = now
    for i in range(need + 1):
        cnt[index] = i
        dfs(index + 1, need - i, buf)
        cnt[index] = 0
        buf = buf * 10 + index


def S(a):
    global ans
    ans = 0
    dfs(0, a, 0)
    return ans


if __name__ == "__main__":
    init()
    print(S(1))
    print(S(5))
    print(S(18))

File: code/885/solution.cpp
#include <iostream>  /*{{{*/
using namespace std; /*}}}*/
const long long mod = 1123455689;
long long jiecheng[20];
int cnt[10];

long long C(int a, int b) {
    if (a < b)
        return 0;
    return jiecheng[a] / jiecheng[b] / jiecheng[a - b] % mod;
}
long long cal() {
    int sum = 0;
    for (int i = 0; i < 10; i++)
        sum += cnt[i];
    long long ret = 1;
    for (int i = 0; i < 10; i++) {
        ret = ret * C(sum - (i == 0), cnt[i]) % mod;
        sum -= cnt[i];
    }
    return ret;
}
void dfs(long long index, long long need, long long now, long long &ans) {
    if (index == 10) {
        ans = (ans + now * cal()) % mod;
        return;
    }
    long long buf = now;
    for (int i = 0; i <= need; i++) {
        cnt[index] = i;
        dfs(index + 1, need - i, buf, ans);
        cnt[index] = 0;
        buf        = (buf * 10 + index) % mod;
    }
}
long long S(int a) {
    long long ans = 0;
    dfs(0, a, 0, ans);
    return ans;
}
void init() {
    jiecheng[0] = 1;
    for (int i = 1; i < 20; i++)
        jiecheng[i] = jiecheng[i - 1] * i;
}
int main() {
    init();
    cout << S(1) << endl;
    cout << S(5) << endl;
    cout << S(18) << endl;
    return 0;
}
